### PR TITLE
Allow gen expr in decl func def

### PR DESCRIPTION
**Pull request by frmdstryr**
For #392

### CODE REVIEW

File: tests/test_inner_functions.py [hunk 1]
```diff
@@ -200,6 +200,17 @@ def test_set_comprehension_func(value):
     assert win.call()
 
 
+# Test that we can access the non local values and the closure variables
+@pytest.mark.parametrize('value', ['self', 'test'])
+def test_generator_expression_func(value):
+    """Test running a generator expression in a declarative function.
+
+    """
+    source = FUNCTION_TEMPLATE.format('any(i for i in range(10))')
```
Comment (MatthieuDartiailh) [inline tests/test_inner_functions.py]: You need to use value like in other tests otherwise you do not test access to non local values and closures.

**Comment by MatthieuDartiailh** [comment]
It may be worth running something similar to the test but outside pytest to avoid mixing things. The Dynamicscope issue seems only related to how pytest try to report the issue. I agree it is worth fixing Dynamicscope though.

**Comment by frmdstryr** [comment]
With #394 the error becomes

```
value = 'self'

    @pytest.mark.parametrize('value', ['self', 'test'])
    def test_generator_expression_func(value):
        """Test running a generator expression in a declarative function.
    
        """
        source = FUNCTION_TEMPLATE.format('any(%s for i in range(10))' % value)
        win = compile_source(source, 'Main')()
>       assert win.call()

tests/test_inner_functions.py:211: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<test>:6: in call
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <range_iterator object at 0x7ff9d111e810>

>   ???
E   NameError: name 'self' is not defined

```

It passes for `test`. The strange trace makes me think I something wrong with the DynamicScope.get.

**Comment by MatthieuDartiailh** [comment]
Actually I also remembered another issue. Gen expr can also be simply returned from the function and for them to behave properly they would need to carry a reference to the Dynamicscope which is why they are not allowed. I don't feel confortable lifting that limitation actually. We could improve the documentation though.

**Comment by frmdstryr** [comment]
Makes sense. So any "closure" that doesn't copy everything before leaving the scope is potentially an issue.

I guess it wouldn't make sense to force evaluation in this case? Eg if the funcdef contains a GenExpr add a decorator that checks if it returned a generator and if so convert it to a tuple?

**Comment by MatthieuDartiailh** [comment]
Honestly I don't see the win in allowing GenExpr would bring us (knowing that list comprehension do work). And your idea would min the parser would need to change behavior based on the presence of the decorator which is not something I like (on top of the painful implementation).

I will close. Feel free to re-open if you feel the need.

**Pull request closed without merging**